cohttp-async/src/io.ml

```diff
@@ -102,6 +102,8 @@ module IO = struct
         Writer.write oc buf;
         return ())
 
+  let refill _ = failwith "Not implemented yet"
+  let with_input_buffer _ = failwith "not implemented yet"
```

Comment:
I'm thinking about how to implement this, and I think it would actually make things easier if `Byte_buffer` was Bigstring based or at least functorized on the buffer type. Reason being is that I don't see how to implement these functions in an efficient manner for Async's Reader for example. I'd imagine we want to use `read_one_chunk_at_a_time` to implement `with_input_buffer`. I understand how bigstrings don't are a wash performance wise, but it helps keep compatibility with existing implementations. Perhaps we could functorize Byte_buffer to make it possible to continue experimenting in Shuttle/Fiber without normal Bytes.

Reply:
I agree but if we want to keep it simple we could check how big of a performace hit would be to just use bigstring

Reply:
Moving to bigstrings will also mean moving the parser to bigstrings. Otherwise we pay the price of allocating more strings?

I was thinking of implementing the async backend in a similar manner as the lwt where there is a byte buffer paired with the reader.

That said, now that I think about it there is going to be some breaking changes in the sense that people using the expert response in servers will need to adapt to the switch to the new input type in the IO backends.

Reply:
> Moving to bigstrings will also mean moving the parser to bigstrings. Otherwise we pay the price of allocating more strings?

Yes but that is very localized to the `Source` module, no? In any case it was just a curiosity, and as you mentioned you also already have an idea to make this a non-issue

> That said, now that I think about it there is going to be some breaking changes in the sense that people using the expert response in servers will need to adapt to the switch to the new input type in the IO backends.

That is the last of the problems. We are allowed to break things in cohttp now, as long as we can justify it. Imho the speedup and latency numbers shown here would justify the breakage

Reply:
`string` is the way to go in OCaml 5.0, since it's both non-moving and allocated via malloc (all large ocaml objects are mallocated in multicore)

Reply:
@avsm That's good to know! With that info I think keeping the API within cohttp to use strings/bytes seems like the way to go. 

> @mseri  That is the last of the problems. We are allowed to break things in cohttp now, as long as we can justify it. Imho the speedup and latency numbers shown here would justify the breakage

👍🏼 That sounds good. And I believe if the libraries using the expert response reference the IO layers by using the IO modules within `cohttp-async`, `cohttp-lwt` etc then we might not even break anything as long as the IO signature stays compatible!

Reply:
> That is the last of the problems. We are allowed to break things in cohttp now, as long as we can justify it. Imho the speedup and latency numbers shown here would justify the breakage

My point is not to avoid breakage, but the observation that we don't have a convenient replacement to offer to the user. Users have a ton of code written using `Writer.t` and `Lwt_io.t` and removing support for these fundamentals is a loss feature wise.

But I think we can still keep normal bytes around. We should just make it possible for users to use Lwt_io/Writer on top even if it means copying. Hopefully, the cost will only be there if these streaming options are used.

Reply:
Fair point!

Reply:
We can try an implementation on top of reader and Lwt_io without an additional buffer. With Async.Reader we will have a way to peek inside the reader's buffer and forward a slice to the parser. I think Lwt_io has something similar via `direct_access`. But I'm not sure this will work for mirage as the mirage-channel api doesn't offer something similar. So we'll need buffering at-least for that backend.

Reply:
We can probably deal with this in another manner by pushing the buffer to the core `cohttp` library, and then the refill becomes a way for the IO layers to fill the buffer managed by cohttp. That will let us keep the IO types for the IO layers the same as they are today, while still benefiting from the buffering we need to drive the parser.

Reply:
I'm not a huge fan of moving a buffer to the core library as that removes one benefit for libraries like shuttle that could otherwise drive the whole cohttp server machinery with just the 1 buffer that's used by its channel implementation.

Reply:
> That will let us keep the IO types for the IO layers the same as they are today, while still benefiting from the buffering we need to drive the parser.

Note that we don't necessarily need to retain the same IO types. For example, if we could just create Reader.t/Writer.t on the fly when the user provides Expert that would be good enough. Though I'm not sure if it's at all possible.

Reply:
> I'm not a huge fan of moving a buffer to the core library as that removes one benefit for libraries like shuttle that could otherwise drive the whole cohttp server machinery with just the 1 buffer that's used by its channel implementation.

Yeah, it's not worth it. If we must break the current api and make Reader/Writer impossible to use, then so be it.

Reply:
`f we could just create Reader.t/Writer.t on the fly`
This should be doable. There will be some awkwardness for sure, but we can definitely get something working. We only need to recreate the reader since I haven't switched the writer in the existing implementation. pseudo code for how I'd do this in async

```ocaml
let reader_of_in_channel t =
   let unconsumed_payload = Bytebuffer.unconsumed_data t.buf in
   let reader_pipe = Reader.pipe t.reader in
   Unix.pipe ~info:(Info.create ...)
  >>| fun (`Reader rd, `Writer wr) ->
        let new_reader = Reader.create rd in
        let writer = Writer.create wr in
       Writer.write writer unconsumed_payload;
        don't_wait_for (
           Writer.transfer writer reader_pipe ~stop:(Reader.close_finished new_reader) (fun s -> Writer.write writer s)
           >>= fun () -> Writer.close writer );
      new_reader
```